### PR TITLE
[dynamo][torch_np] support linalg, random and fft module

### LINKED ISSUE #105269: graph break in compiling `np.linalg.norm` and other np.linalg functions
### 🐛 Describe the bug

```
$ cat compile_norm.py 
import numpy as np
import matplotlib.pyplot as plt

import torch
torch.set_default_device("cpu")
import torch._dynamo.config as cfg
cfg.numpy_ndarray_as_tensor = True

@torch.compile
def func(x):
    return np.linalg.norm(x)

print(func([1, 2, 3]))
```

Running  `$ TORCH_LOGS="+dynamo" python compile_norm.py`

produces

```
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /home/br/mambaforge/envs/torch_nightly/lib/python3.8/contextlib.py
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /home/br/mambaforge/envs/torch_nightly/lib/python3.8/contextlib.py
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping helper /home/br/mambaforge/envs/torch_nightly/lib/python3.8/contextlib.py
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /home/br/mambaforge/envs/torch_nightly/lib/python3.8/contextlib.py
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /home/br/mambaforge/envs/torch_nightly/lib/python3.8/contextlib.py
[2023-07-15 15:10:04,923] torch._dynamo.eval_frame: [DEBUG] skipping enable_dynamic /home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py
[2023-07-15 15:10:04,923] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo start tracing func compile_norm.py:10
[2023-07-15 15:10:04,925] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line compile_norm.py:12
[2023-07-15 15:10:04,925] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL np []
[2023-07-15 15:10:04,925] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR linalg [PythonModuleVariable()]
[2023-07-15 15:10:04,926] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR norm [PythonModuleVariable()]
[2023-07-15 15:10:04,926] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST x [SkipFilesVariable()]
[2023-07-15 15:10:04,926] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 1 [SkipFilesVariable(), ListVariable()]
[2023-07-15 15:10:04,927] torch._dynamo.symbolic_convert: [DEBUG] break_graph_if_unsupported triggered compile
Traceback (most recent call last):
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 391, in wrapper
    return inner_fn(self, inst)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1100, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 559, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/variables/misc.py", line 765, in call_function
    unimplemented(
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 141, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_function norm in skip_files <__array_function__ internals>
[2023-07-15 15:10:04,927] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: call_function norm in skip_files <__array_function__ internals> from user code at   File "compile_norm.py", line 12, in func
    return np.linalg.norm(x)

[2023-07-15 15:10:04,927] torch._dynamo.output_graph: [DEBUG] restore_graphstate: removed 0 nodes
[2023-07-15 15:10:04,927] torch._dynamo.output_graph: [DEBUG] COMPILING GRAPH due to GraphCompileReason(reason='call_function norm in skip_files <__array_function__ internals>', user_stack=[<FrameSummary file compile_norm.py, line 12 in func>], graph_break=True)
[2023-07-15 15:10:04,943] torch._dynamo.convert_frame.__guards: [DEBUG] GUARDS:
  ___check_type_id(L['x'], 94580818297504)
  len(L['x']) == 3
  ___check_type_id(L['x'][0], 94580818295744)
  L['x'][0] == 1
  ___check_type_id(L['x'][1], 94580818295744)
  L['x'][1] == 2
  ___check_type_id(L['x'][2], 94580818295744)
  L['x'][2] == 3
  ___is_grad_enabled()
  not ___are_deterministic_algorithms_enabled()
  utils_device.CURRENT_DEVICE == device(type='cpu')
[2023-07-15 15:10:04,943] torch._dynamo.eval_frame: [DEBUG] skipping norm <__array_function__ internals>
[2023-07-15 15:10:04,943] torch._dynamo.eval_frame: [DEBUG] skipping _norm_dispatcher /home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/numpy/linalg/linalg.py
[2023-07-15 15:10:04,943] torch._dynamo.eval_frame: [DEBUG] skipping norm /home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/numpy/linalg/linalg.py
[2023-07-15 15:10:04,944] torch._dynamo.eval_frame: [DEBUG] skipping isComplexType /home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/numpy/linalg/linalg.py
3.7416573867739413
[2023-07-15 15:10:04,948] torch._dynamo.utils: [INFO] TorchDynamo compilation metrics:
Function      Runtimes (s)
----------  --------------
_compile            0.0197
```

Same error for other functions from `numpy.linalg` namespace (solve etc).

### Versions

same as https://github.com/pytorch/pytorch/issues/105052

cc @mruberry @rgommers @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### PR DESCRIPTION
**Pull request by kshitij12345**
Support tracing through `np.linalg` with `torch_np` installed. Will update with other modules if this approach makes sense.

TODO:
* [x] Add test for `fft` and `random`.

Fixes https://github.com/pytorch/pytorch/issues/105269

cc @mruberry @rgommers @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/allowed_functions.py
Comment (kshitij12345) [inline torch/_dynamo/allowed_functions.py:256]: Not sure about the implication of supporting np.linalg functions when `HAS_NUMPY=True and HAS_NUMPY_TORCH_INTEROP=False`, so being conservative here.

File: torch/_dynamo/variables/misc.py
Comment (kshitij12345) [inline torch/_dynamo/variables/misc.py:835]: Is there a better way to do this?

**Comment by kshitij12345** [comment]
cc: @lezcano

File: test/dynamo/test_functions.py
Comment (lezcano) [inline test/dynamo/test_functions.py:1133]: No need of using the `from_numpy`. I added a patch so that the test without `from_numpy` passes in https://github.com/pytorch/pytorch/pull/105333.

**Review by lezcano — COMMENTED** [review]
This LGTM, but let's wait for a proper review from a dynamo maintainer.

File: torch/_dynamo/variables/misc.py
Comment (larryliu0820) [reply torch/_dynamo/variables/misc.py:835]: In addition to that, can we extract this logic into a separate method? It's becoming more complicated. Would be good to have this util that takes in `self.value` and emits `func`, or throws proper exception.

Comment (larryliu0820) [reply torch/_dynamo/variables/misc.py:835]: Also cc @voznesenskym

File: torch/_dynamo/allowed_functions.py
Comment (ev-br) [reply torch/_dynamo/allowed_functions.py:256]: np.fft, too?

Comment (lezcano) [reply torch/_dynamo/allowed_functions.py:256]: yep, but see the OP

Comment (ezyang) [reply torch/_dynamo/allowed_functions.py:256]: Yeah, I'd try increasing modules unconditionally

File: torch/_dynamo/variables/misc.py
Comment (ezyang) [reply torch/_dynamo/variables/misc.py:835]: Iterating here seems a bit wacky to me. Why is this not just a straight dict lookup?

Comment (lezcano) [reply torch/_dynamo/variables/misc.py]: On a second thought, the module should be somewhere. Otherwise we couldn't tell apart, say, `torch.norm` and `torch.linalg.norm`!

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py]: At this point, `self.value` is a numpy function and numpy doesn't seem to consistently have `__module__` on all the functions (seems like `ufuncs` don't have `__module__`) set. (Probably a numpy bug)

```python
>>> import numpy as np
>>> np.sin.__module__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'numpy.ufunc' object has no attribute '__module__'
>>> np.mean
<function mean at 0x7f067bf571f0>
>>> np.mean.__module__
'numpy'
```

So, I don't think we can rely on it.

Comment (lezcano) [reply torch/_dynamo/variables/misc.py]: what about trying to get the module, and:
- If it has it, then fetch it from the correct module
- If it doesn't have it, then fetch the function from the main `np` module
?

I reckon that functions that live in a NumPy module do have their `__module__` attribute correctly set.

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py]: Just found that `np.random.randn.__module__` returns `None`. I don't think information from `__module__` is reliable.

So, we need something like
```python
def get_torch_np_func(fn):
    import torch_np

    try:
        module = fn.__module__
        if not module:
            # Not sure if other submodules also do this.
            module = 'numpy.random'
    except AttributeError:
        module = 'numpy'

    module = module.replace('numpy', 'torch_np')
    module_qualified_name = module + "." + fn.__name__
    obj = torch_np

    try:
        # first is always `torch_np`, so we skip it.
        for path in module_qualified_name.split(".")[1:]:
            obj = getattr(obj, path)
        return obj
    except AttributeError:
        unimplemented(f"Can't find numpy function {fn} in torch_np")

```

Comment (lezcano) [reply torch/_dynamo/variables/misc.py]: This HORRIBLE implementation is able to get the module of all the numpy functions we implement (and more):
```python
def get_np_module(fn):
    if isinstance(fn, ufunc):
        return "numpy"
    module = getattr(fn, "__module__", None)
    if module is None:
        s = str(fn)
        s = s.split(" ")
        # Handles objects like
        # <built-in method randn of numpy.random.mtrand.RandomState object at 0x7ff67319a940>
        if len(s) == 8:
            s = s[4]
            s = s.split(".")
            if s[0] == "numpy":
                module = f"{s[0]}.{s[1]}"
    elif "numpy.random" in module:
        # Handles np.random.MT19937.__module__ == 'numpy.random._mt19937.MT19937'
        # There's many of this form in torch.randn
        module = "numpy.random"
    return module
```

The test:
```python
def get_fns(mod):
    for fn in dir(mod):
        if fn[0] != "_":
            f = getattr(mod, fn)
            if callable(f):
                yield f


def test_module(mod):
    for fn in get_fns(mod):
        if (module := get_np_module(fn)) != mod.__name__:
            print(fn, module)



import numpy as np
test_module(np)
import numpy.fft as fft
test_module(fft)
import numpy.linalg as linalg
test_module(linalg)
import numpy.random as random
test_module(random)
```

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py]: I was planning to create this map once and use it later for indexing.
```python
    from ..utils import NP_SUPPORTED_MODULES, NP_TO_TORCH_NP_MODULE

    # NP_SUPPORTED_MODULES = (np, np.fft, np.linalg, np.random)
    # NP_TO_TORCH_NP_MAP = {
    #    np: torch_np,
    #    np.fft: torch_np.fft,
    #    np.linalg: torch_np.linalg,
    #    np.random: torch_np.random,
    # }

    def get_np_to_torch_np_map():
        np_fn_to_torch_np_fn = {}

        for mod in NP_SUPPORTED_MODULES:
            np_fn_to_torch_np_fn.update(
                {
                    # `torch_np` doesn't implement functions like
                    # like `numpy.fromstring`, `numpy.fromfile`, etc.
                    v: getattr(NP_TO_TORCH_NP_MODULE[mod], k, None)
                    for k, v in mod.__dict__.items()
                    if callable(v)
                    and (getattr(v, "__module__", None) or mod.__name__) == mod.__name__
                }
            )

        return np_fn_to_torch_np_fn

    NP_FN_TO_TORCH_NP_FN = get_np_to_torch_np_map()
```

**Review by ezyang — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
Just one note, otherwise it LGTM.

Thank you!

File: torch/_dynamo/variables/misc.py
Comment (lezcano) [reply torch/_dynamo/variables/misc.py]: that's certainly better, yep.

**Pull request closed without merging**
